### ISSUE TITLE
prediction outputs & model classes

### ISSUE THREAD

**Issue by hkmztrk** [jerryji1993/DNABERT#10]
Hello,

**1. Prediction outputs** 

When I run fine-tuning script on prom-core set, I get the following results:

```
01/12/2021 16:02:40 - INFO - __main__ -   ***** Eval results  *****
01/12/2021 16:02:40 - INFO - __main__ -     acc = 0.49248183814833585
01/12/2021 16:02:40 - INFO - __main__ -     auc = 0.5877368768457378
01/12/2021 16:02:40 - INFO - __main__ -     aupr = 0.5465873116339481
01/12/2021 16:02:40 - INFO - __main__ -     f1 = 0.6599501924383065
01/12/2021 16:02:40 - INFO - __main__ -     mcc = 0.0
01/12/2021 16:02:40 - INFO - __main__ -     precision = 0.49248183814833585
01/12/2021 16:02:40 - INFO - __main__ -     recall = 1.0
```

After that running the prediction script, the  results are:

```
01/12/2021 16:47:11 - INFO - __main__ -   ***** Pred results  *****
01/12/2021 16:47:11 - INFO - __main__ -     acc = 0.49248183814833585
01/12/2021 16:47:11 - INFO - __main__ -     auc = 0.587710896620401
01/12/2021 16:47:11 - INFO - __main__ -     aupr = 0.5466122449761592
01/12/2021 16:47:11 - INFO - __main__ -     f1 = 0.6599501924383065
01/12/2021 16:47:11 - INFO - __main__ -     mcc = 0.0
01/12/2021 16:47:11 - INFO - __main__ -     precision = 0.49248183814833585
01/12/2021 16:47:11 - INFO - __main__ -     recall = 1.0
```

The scores for two different tasks are almost almost completely identical. Also, not sure which dataset in the paper  'prom-core' corresponds to (maybe general), but the scores here looks quite different from the ones in paper. 

Finally, when I ran fine-tune script for my own dataset, the results look like this

```
01/11/2021 19:39:51 - INFO - __main__ -   ***** Eval results  *****
01/11/2021 19:39:51 - INFO - __main__ -     acc = 0.5020841847159787
01/11/2021 19:39:51 - INFO - __main__ -     auc = 0.4945063291829262
01/11/2021 19:39:51 - INFO - __main__ -     aupr = 0.4952829066522459
01/11/2021 19:39:51 - INFO - __main__ -     f1 = 0.0
01/11/2021 19:39:51 - INFO - __main__ -     mcc = 0.0
01/11/2021 19:39:51 - INFO - __main__ -     precision = 0.0
01/11/2021 19:39:51 - INFO - __main__ -     recall = 0.0
```

I'm wondering why the last four metrics are zero? I am sure the data is formatted similar to your example, otherwise it wouldn't be able train in the first place, right? What could be the problem?

**2. Different model classes**

In `run_pretrain.py `there are several classes including Roberta and GPT. Can we also use these for pretraining if we use `DNATokenizer` ? 

I'd appreciate it a lot if you could give some information about these two questions.

**Comment by Zhihan1996** [comment]
Hi,

For the first question, could you please show me the command you are using when running the model. 

For the second question, yes, you can use it. But in our case, BERT and Roberta essential the same. GPT is for sequence generation. So I am not sure if it makes sense to use it in the DNA setting.

**Comment by Z-Abbas** [comment]
Hi
I have a similar issue. I run the example using the commands below:

3.3 Fine-tune with pre-trained model
export KMER=6
export MODEL_PATH='/home/zeeshan/DNABERT/6-new-12w-0/'
export DATA_PATH='/home/zeeshan/DNABERT/examples/sample_data/ft/prom-core/6/'
export OUTPUT_PATH='/home/zeeshan/DNABERT/examples/OUTPUT/'

4. Prediction
export KMER=6
export MODEL_PATH='/home/zeeshan/DNABERT/examples/OUTPUT/'
export DATA_PATH='/home/zeeshan/DNABERT/examples/sample_data/ft/prom-core/6/'
export PREDICTION_PATH='/home/zeeshan/DNABERT/examples/predout/'

After running the training and prediction codes, the result of the last (100th) one is exactly the same as the prediction result. Am I doing something wrong? or is there any option to choose the best weights for prediction purpose?
Please, find the eval_results.txt file attached with the screenshot of the prediction result for your reference.

[eval_results.txt](https://github.com/jerryji1993/DNABERT/files/6225616/eval_results.txt)
![pred_results](https://user-images.githubusercontent.com/80881943/112921518-e115b800-9145-11eb-94ca-fff73b5dd0cf.PNG)

**Comment by nathanhu3** [comment]
I'm having similar issues. I am following the readme instructions and getting poor evaluation and prediction results after fine tuning. I downloaded the DNABERT6 pre-trained model and ran the fine-tune command for the `prom-core` task under section 3.3. 

I get similar behaviour as seen in @Z-Abbas' `eval_results.txt` where the fine-tuning will be running okay until all of a sudden the accuracy drops to 0.507 or 0.492 and is stuck there, sometimes until the end of fine-tuning, causing the final prediction accuracy to be around the same. Sometimes this behaviour will stop after about 1000-2000 steps and if it does the final prediction accuracy might be a bit better (0.7-0.85). After running fine-tuning from a fresh clone several times, the best accuracy I can achieve is ~0.85 but most attempts either finish fine-tuning with accuracies of exactly 0.507/0.492 or around 0.75. I am using the repo exactly (using the same `dev.tsv` and `train.tsv`). I have also noticed that the loss often doesn't improve over the 3 epochs after the first few hundred steps.

TLDR; Results from fine-tuning DNABERT several times yields poor accuracies. Often it is either exactly 0.492 or 0.507, while other times it will be between 0.7-0.85.

Thanks for your help.

**Comment by nathanhu3** [comment]
I'm having similar issues. I am following the readme instructions and getting poor evaluation and prediction results after fine tuning. I downloaded the DNABERT6 pre-trained model and ran the fine-tune command for the `prom-core` task under section 3.3. 

I get similar behaviour as seen in @Z-Abbas' `eval_results.txt` where the fine-tuning will be running okay until all of a sudden the accuracy drops to 0.507 or 0.492 and is stuck there, sometimes until the end of fine-tuning, causing the final prediction accuracy to be around the same. Sometimes this behaviour will stop after about 1000-2000 steps and if it does the final prediction accuracy might be a bit better (0.7-0.85). After running fine-tuning from a fresh clone several times, the best accuracy I can achieve is ~0.85 but most attempts either finish fine-tuning with accuracies of exactly 0.507/0.492 or around 0.75. I am using the repo exactly (using the same `dev.tsv` and `train.tsv`). I have also noticed that the loss often doesn't improve over the 3 epochs after the first few hundred steps.

TLDR; Results from fine-tuning DNABERT several times yields poor accuracies. Often it is either exactly 0.492 or 0.507, while other times it will be between 0.7-0.85.

Thanks for your help.

**Comment by Zhihan1996** [comment]
Hi,

I think this task is relatively tricky. We found that the model may fail to converge in some cases. So please try different hyperparameter settings and random seed.

**Comment by woreom** [comment]
> Hi,
> 
> I think this task is relatively tricky. We found that the model may fail to converge in some cases. So please try different hyperparameter settings and random seed.

Can you give us an example? because personally it a bit hard for me to follow your codes, even for simplest things like loading the model and pridicting labels

**Comment by NikitaBhandare** [comment]
@jerryji1993 I am still facing the same issue even after cloning the latest repo. Can you please advice as I have my research based on this model.
